src/EditorFeatures/Core/Implementation/EncapsulateField/AbstractEncapsulateFieldCommandHandler.cs

```diff
@@ -84,23 +83,30 @@ private bool Execute(EncapsulateFieldCommandArgs args, IUIThreadOperationScope w
                 var service = document.GetLanguageService<AbstractEncapsulateFieldService>();
 
                 var result = service.EncapsulateFieldAsync(document, spans.First().Span.ToTextSpan(), true, cancellationToken).WaitAndGetResult(cancellationToken);
+                var finalSolution = result?.GetSolutionAsync(cancellationToken).WaitAndGetResult(cancellationToken);
+
+                // This is the last point where the operation can be canceled by the operation context
+                // managed by the command system. Make sure to not ignore a cancellation request which
+                // occurred prior to this line before proceeding with the Roslyn-managed dialogs.
+                cancellationToken.ThrowIfCancellationRequested();
 
                 // We are about to show a modal UI dialog so we should take over the command execution
                 // wait context. That means the command system won't attempt to show its own wait dialog 
                 // and also will take it into consideration when measuring command handling duration.
                 waitScope.Context.TakeOwnership();
 
+                // The cancellation token associated with the operation scope may or may not have been
+                // canceled when ownership of the context was taken. Treat the token as invalid and use
+                // a different one for future operations in this command handler.
+                cancellationToken = CancellationToken.None;
+
                 if (result == null)
                 {
                     var notificationService = workspace.Services.GetService<INotificationService>();
                     notificationService.SendNotification(EditorFeaturesResources.Please_select_the_definition_of_the_field_to_encapsulate, severity: NotificationSeverity.Error);
                     return false;
                 }
 
-                waitScope.AllowCancellation = false;
-
-                var finalSolution = result.GetSolutionAsync(cancellationToken).WaitAndGetResult(cancellationToken);
-
                 var previewService = workspace.Services.GetService<IPreviewDialogService>();
                 if (previewService != null)
```

Comment:
[pretend this comment is on line 117]

The calls there are still using the cancellationToken, which now "looks" like a bug. Your reassigning of it means it's _not_ a bug, but this still looks very smelly/incomplete. I'm with @CyrusNajmabadi's suggestion to just move that work up, then this "appears" more complete without a careful analysis.